src/Streamly/Internal/Data/Parser/Types.hs

```diff
@@ -181,7 +252,220 @@ splitWith func (Parser stepL initialL extractL)
+            Error _ -> do
+                xs <- fextract fs
+                return $ Stop cnt1 xs
+
+    -- XXX The "try" may impact performance if this parser is used as a scan
+    extract (Tuple3' s _ fs) = do
+        r <- try $ extract1 s
+        case r of
+            Left (_ :: ParseError) -> fextract fs
+            Right b -> fstep fs b >>= fextract
+
+-- | See documentation of 'Streamly.Internal.Data.Parser.some'.
+--
+-- /Internal/
+--
+{-# INLINE splitSome #-}
+splitSome :: MonadCatch m => Fold m b c -> Parser m a b -> Parser m a c
+splitSome (Fold fstep finitial fextract) (Parser step1 initial1 extract1) =
+    Parser step initial extract
+
+    where
+
+    initial = do
+        ps <- initial1 -- parse state
+        fs <- finitial -- fold state
+        pure (Tuple3' ps 0 (Left fs))
+
+    {-# INLINE step #-}
+    step (Tuple3' st _ (Left fs)) a = do
+        r <- step1 st a
+        case r of
+            Yield _ s -> return $ Skip 0 (Tuple3' s undefined (Left fs))
+            Skip  n s -> return $ Skip n (Tuple3' s undefined (Left fs))
+            Stop n b -> do
+                s <- initial1
+                fs1 <- fstep fs b
+                -- XXX this is also a yield point, we will never fail beyond
+                -- this point. If we do not yield then if an error occurs after
+                -- this then we will backtrack to the previous yield point
+                -- instead of this point which is wrong.
+                --
+                -- so we need a yield with backtrack
+                return $ Skip n (Tuple3' s 0 (Right fs1))
+            Error err -> return $ Error err
+    step (Tuple3' st cnt (Right fs)) a = do
+        r <- step1 st a
+        let cnt1 = cnt + 1
+        case r of
+            Yield _ s -> return $ Yield 0 (Tuple3' s cnt1 (Right fs))
+            Skip n s -> do
+                assert (cnt1 - n >= 0) (return ())
+                return $ Skip n (Tuple3' s (cnt1 - n) (Right fs))
+            Stop n b -> do
+                s <- initial1
+                fs1 <- fstep fs b
+                -- XXX we need to yield here but also backtrack
+                return $ Skip n (Tuple3' s 0 (Right fs1))
+            Error _ -> Stop cnt1 <$> fextract fs
+
+    -- XXX The "try" may impact performance if this parser is used as a scan
+    extract (Tuple3' s _ (Left fs)) = extract1 s >>= fstep fs >>= fextract
+    extract (Tuple3' s _ (Right fs)) = do
+        r <- try $ extract1 s
+        case r of
+            Left (_ :: ParseError) -> fextract fs
+            Right b -> fstep fs b >>= fextract
+
+-- This is the dual of "nil".
+--
+-- | A parser that always fails with an error message without consuming
+-- any input.
+--
+-- /Internal/
+--
+{-# INLINE die #-}
+die :: Applicative m => String -> Parser m a b
+die err =
+    Parser (\_ _ -> pure $ Error err) -- step
+           (pure ())                  -- initial
+           (\_ -> error err)          -- extract
```

Comment:
`throwM` instead of `error`?

Reply:
Yes, this code is from before I introduced ParseError, missed changing it.